crates/red_knot_python_semantic/src/types.rs

```diff
@@ -5800,18 +5863,55 @@ impl<'db> FunctionType<'db> {
     /// Were this not a salsa query, then the calling query
     /// would depend on the function's AST and rerun for every change in that file.
     #[salsa::tracked(return_ref)]
-    pub(crate) fn signature(self, db: &'db dyn Db) -> Signature<'db> {
+    pub(crate) fn signature(self, db: &'db dyn Db) -> FunctionSignature<'db> {
         let mut internal_signature = self.internal_signature(db);
 
-        if self.has_known_decorator(db, FunctionDecorators::OVERLOAD) {
-            return Signature::todo("return type of overloaded function");
+        if let Some(specialization) = self.specialization(db) {
+            internal_signature = internal_signature.apply_specialization(db, specialization);
         }
 
-        if let Some(specialization) = self.specialization(db) {
-            internal_signature.apply_specialization(db, specialization);
+        // The semantic model records a use for each function on the name node. This is used here
+        // to get the previous function definition with the same name.
+        let scope = self.definition(db).scope(db);
+        let use_def = semantic_index(db, scope.file(db)).use_def_map(scope.file_scope_id(db));
+        let use_id = self
+            .body_scope(db)
+            .node(db)
+            .expect_function()
+            .name
+            .scoped_use_id(db, scope);
+
+        if let Symbol::Type(Type::FunctionLiteral(function_literal), Boundness::Bound) =
```

Comment:
We could insert a comment here
```suggestion
        // TODO maybe someday: infer union of overloaded callables if we have a maybe-bound overload?
        if let Symbol::Type(Type::FunctionLiteral(function_literal), Boundness::Bound) =
```

Reply:
Hmm, interesting. Would you be able to provide an example of where this might be possible?

Would it be something like this:
```py
def _(flag: bool) -> None:
    if flag:

        @overload
        def foo() -> None: ...
        @overload
        def foo(x: int) -> int: ...

	# Here, the previous `foo` overloads are possibly-unbound
    @overload
    def foo(x: str) -> str: ...
    @overload
    def foo(x: bytes) -> bytes: ...
    def foo(x: int | str | bytes | None = None) -> int | str | bytes | None:
        return x

    reveal_type(foo)
```

This actually prompted me to add some test cases to include an overload in a union.